src/shared/sam/cli/samCliProcess.ts

```diff
@@ -0,0 +1,94 @@
+/*!
+ * Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+'use strict'
+
+import * as child_process from 'child_process'
+import * as events from 'events'
+
+export interface SamCliProcessResult {
+    process: SamCliProcess,
+    exitCode: number,
+    error: Error | undefined,
+    stdout: string,
+    stderr: string
+}
+
+/**
+ * Manages a child process making a sam cli call
+ * To use:
+ * - instantiate
+ * - call start
+ * - await promise to get the results
+ */
+export class SamCliProcess {
+    private static readonly CHILD_PROCESS_CLOSED = 'childProcessClosed'
+
+    private readonly _process: string
+    private readonly _args: string[] | undefined
+
+    private _childProcess: child_process.ChildProcess | undefined = undefined
+    private readonly _onChildProcessClosed: events.EventEmitter = new events.EventEmitter()
+    private readonly _stdoutChunks: string[] = []
+    private readonly _stderrChunks: string[] = []
+    private _error: Error | undefined
+    private readonly _processCompletedPromise: Promise<SamCliProcessResult>
+
+    public constructor(process: string, args?: string[] | undefined) {
+        this._process = process
+        this._args = args
+
+        this._processCompletedPromise = new Promise((resolve, reject) => {
+            this._onChildProcessClosed.once(
+                SamCliProcess.CHILD_PROCESS_CLOSED,
+                (processResult: SamCliProcessResult) => {
```

Comment:
What if `processResult` has a non-0 exit code? Should we `reject` instead?

Reply:
Interesting idea. I think that's worth exploring for sam cli. We would still reject the payload yes?

Reply:
Instead I will move the non 0 exit check to samCliCommand with the revamp you suggested